rust/src/tx_builder.rs

```diff
@@ -265,118 +271,252 @@ impl TransactionBuilder {
-                    while added < remaining_amount {
-                        if available_inputs.is_empty() {
-                            return Err(JsError::from_str("UTxO Balance Insufficient"));
-                        }
-                        let (new_input_total, new_output_total, new_added, new_remaining_amount, _) = add_random_input(self, &mut rng, &mut available_inputs, &input_total, &output_total, &added, &remaining_amount)?;
-                        input_total = new_input_total;
-                        output_total = new_output_total;
-                        added = new_added;
-                        remaining_amount = new_remaining_amount;
+                while input_total.coin < output_total.coin {
+                    if available_indices.is_empty() {
+                        return Err(JsError::from_str("UTxO Balance Insufficient[x]"));
                     }
+                    let i = *available_indices.iter().nth(rng.gen_range(0..available_indices.len())).unwrap();
+                    available_indices.remove(&i);
+                    let input = &available_inputs[i];
+                    let input_fee = self.fee_for_input(&input.output.address, &input.input, &input.output.amount)?;
+                    self.add_input(&input.output.address, &input.input, &input.output.amount);
+                    input_total = input_total.checked_add(&input.output.amount)?;
+                    output_total = output_total.checked_add(&Value::new(&input_fee))?;
                 }
             },
         }
 
         Ok(())
     }
 
+    fn cip2_largest_first_by<F>(
+        &mut self,
+        available_inputs: &Vec<TransactionUnspentOutput>,
+        available_indices: &mut Vec<usize>,
+        input_total: &mut Value,
+        output_total: &mut Value,
+        by: F) -> Result<(), JsError>
+    where
+        F: Fn(&Value) -> Option<BigNum> {
+        let mut relevant_indices = available_indices.clone();
+        relevant_indices.retain(|i| by(&available_inputs[*i].output.amount).is_some());
+        // ordered in ascending order by predicate {by}
+        relevant_indices.sort_by_key(|i| by(&available_inputs[*i].output.amount).expect("filtered above"));
+
+        // iterate in decreasing order for predicate {by}
+        for i in relevant_indices.iter().rev() {
+            if by(input_total).unwrap_or(BigNum::zero()) >= by(output_total).expect("do not call on asset types that aren't in the output") {
+                break;
+            }
+            let input = &available_inputs[*i];
+            // differing from CIP2, we include the needed fees in the targets instead of just output values
+            let input_fee = self.fee_for_input(&input.output.address, &input.input, &input.output.amount)?;
+            self.add_input(&input.output.address, &input.input, &input.output.amount);
+            *input_total = input_total.checked_add(&input.output.amount)?;
```

Comment:
Doing `?` here might not return a sensible error since that would be `JsError::from(None)`

Reply:
We plan on going through the builder in the coming while to improve errors so they're more helpful for the user so it's good you're looking at these. Where does `None` come from though? I thought it would return the error type from `BigNum::checked_add()` (that seems to be the only part where `Value::checked_add()` can error) which is `JsError::from_str("overflow")`.

Reply:
Oh, I assumed it was an integer or float type. Sorry about that, I realize it is already a JsError when returned from checked_add.